src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicBuffers.cs

```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Microsoft.Quic;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal;
+
+/// <summary>
+/// Helper class to convert managed data into QUIC_BUFFER* consumable by MsQuic.
+/// It also allows reuse with repeated Reset/Initialize calls.
+/// Note that since this is struct and there's no finalizer, Dispose must be always called to release the unmanaged memory allocated by this struct.
+/// </summary>
+internal unsafe struct MsQuicBuffers : IDisposable
+{
+    // Handles to pinned memory blocks from the user.
+    private MemoryHandle[] _handles;
+    // Native memory block which holds the pinned memory pointers from _handles and can be passed to MsQuic as QUIC_BUFFER*.
+    private void* _buffers;
+    // Number of QUIC_BUFFER instance currently allocated in _buffers, so that we can reuse the memory instead of reallocating.
+    private int _count;
+
+    public MsQuicBuffers()
+    {
+        _handles = Array.Empty<MemoryHandle>();
+        _buffers = null;
+        _count = 0;
+    }
+
+    public QUIC_BUFFER* Buffers => (QUIC_BUFFER*)_buffers;
+    public int Count => _count;
+
+    /// <summary>
+    /// The mothod initializes QUIC_BUFFER* with data from inputs, converted via toBuffer.
+    /// Note that the struct either needs to be freshly created or previously cleaned up with Reset.
+    /// </summary>
+    /// <param name="inputs">Inputs to get their byte array, pin them and pepare them to be passed to MsQuic as QUIC_BUFFER*.</param>
+    /// <param name="toBuffer">Method extracting byte array from the inputs, e.g. applicationProtocol.Protocol.</param>
+    /// <typeparam name="T">The type of the inputs.</typeparam>
+    public void Initialize<T>(IList<T> inputs, Func<T, ReadOnlyMemory<byte>> toBuffer)
+    {
+        if (_handles.Length < inputs.Count)
+        {
+            ArrayPool<MemoryHandle>.Shared.Return(_handles);
+            _handles = ArrayPool<MemoryHandle>.Shared.Rent(inputs.Count);
```

Comment:
Assuming that you are going to implement reusing of the `MemoryHandle` arrays, it is not obvious to me that additional pooling these arrays is worth it. ArrayPool is not "free".

Reply:
`MsQuicBuffers` is allocated per `QuicStream` (e.g. one request/response pair) and reused for repeated send operations. Whereas the pool allows us to "reuse" the `MemoryHandle` arrays across the whole app/Quic lib lifetime.

Reply:
It is typically profitable to pool objects that are big (e.g. byte[] buffers) and/or that are allocated very frequently. Pooling these MemoryHandle arrays across the whole app is neither.

Global pooling is replacing one cost with a different cost. The objects in the pool make the garbage collector work harder. It is not always a win, especially for small objects like what these MemoryHandle[] arrays are typically going to be. Unless you have done a performance measurements that show that the global pooling is improving actual performance (e.g. requests per second), I would not introduce it just because you can.